File: dialogs.go
package main

import (
	"os/exec"
	"strings"
)

// promptForName shows a native text-input dialog. The button picks the
// kind: Chore for things to do again, Streak for things being avoided.
// ok is false when the dialog is cancelled or the input is empty.
func promptForName() (name string, streak, ok bool) {
	// The input is a single line, so a linefeed can't appear in the
	// name and safely separates it from the button choice.
	out, err := exec.Command("osascript", "-e",
		`set d to display dialog "What do you want to track?" default answer "" with title "Since" buttons {"Cancel", "Streak", "Chore"} default button "Chore"`,
		"-e", `(text returned of d) & linefeed & (button returned of d)`,
	).Output()
	if err != nil { // cancelled
		return "", false, false
	}
	text, button, found := strings.Cut(strings.TrimRight(string(out), "\n"), "\n")
	if !found {
		return "", false, false
	}
	name = strings.TrimSpace(text)
	return name, button == "Streak", name != ""
}

// promptForEdit shows one dialog for an item: its name in a text field,
// with Remove and Rename as the actions. button is "Remove" or "Rename";
// ok is false when cancelled.
func promptForEdit(current string) (name, button string, ok bool) {
	out, err := exec.Command("osascript", "-e",
		`set d to display dialog "Edit “`+escapeAppleScript(current)+`”:" default answer "`+escapeAppleScript(current)+`" with title "Since" buttons {"Cancel", "Remove", "Rename"} default button "Rename"`,
		"-e", `(text returned of d) & linefeed & (button returned of d)`,
	).Output()
	if err != nil { // cancelled
		return "", "", false
	}
	text, button, found := strings.Cut(strings.TrimRight(string(out), "\n"), "\n")
	if !found {
		return "", "", false
	}
	return strings.TrimSpace(text), button, true
}

// alertDuplicate tells the user an item with this name already exists.
func alertDuplicate(name string) {
	exec.Command("osascript", "-e",
		`display alert "“`+escapeAppleScript(name)+`” already exists" message "Item names must be unique." as warning`,
	).Run()
}

// confirmEndStreak asks before resetting a streak — a misclick would
// kill the trophy. length is the current streak ("4 months"), or ""
// when it's too young to phrase that way.
func confirmEndStreak(name, length string) bool {
	msg := `End “` + escapeAppleScript(name) + `” streak?`
	if length != "" {
		msg = `End your ` + escapeAppleScript(length) + ` “` + escapeAppleScript(name) + `” streak?`
	}
	err := exec.Command("osascript", "-e",
		`display dialog "`+msg+`" with title "Since" buttons {"Cancel", "End Streak"} default button "Cancel"`,
	).Run()
	return err == nil // cancelled dialogs exit non-zero
}

// confirmRemove shows a native confirm dialog; true means delete it.
func confirmRemove(name string) bool {
	err := exec.Command("osascript", "-e",
		`display dialog "Remove “`+escapeAppleScript(name)+`”?" with title "Since" buttons {"Cancel", "Remove"} default button "Cancel"`,
	).Run()
	return err == nil // cancelled dialogs exit non-zero
}

// escapeAppleScript escapes a value for use inside an AppleScript string literal.
func escapeAppleScript(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	return strings.ReplaceAll(s, `"`, `\"`)
}

File: Sources/Since/Dialogs.swift
import AppKit

/// Which action the edit dialog chose.
enum EditAction {
    case rename
    case remove
}

/// Shows a text-input alert for a new item. The button picks the kind:
/// Chore for things to do again, Streak for things being avoided.
/// Returns nil when cancelled or the input is empty.
func promptForName() -> (name: String, isStreak: Bool)? {
    let alert = NSAlert()
    alert.messageText = "What do you want to track?"
    alert.addButton(withTitle: "Chore")
    alert.addButton(withTitle: "Streak")
    alert.addButton(withTitle: "Cancel")
    let field = textField()
    alert.accessoryView = field
    alert.window.initialFirstResponder = field

    let response = run(alert)
    let name = field.stringValue.trimmingCharacters(in: .whitespaces)
    if name.isEmpty {
        return nil
    }
    switch response {
    case .alertFirstButtonReturn: return (name, false)
    case .alertSecondButtonReturn: return (name, true)
    default: return nil
    }
}

/// Shows one alert for an item: its name in a text field, with Remove and
/// Rename as the actions. Returns nil when cancelled.
func promptForEdit(_ current: String) -> (name: String, action: EditAction)? {
    let alert = NSAlert()
    alert.messageText = "Edit “\(current)”:"
    alert.addButton(withTitle: "Rename")
    alert.addButton(withTitle: "Remove")
    alert.addButton(withTitle: "Cancel")
    let field = textField()
    field.stringValue = current
    alert.accessoryView = field
    alert.window.initialFirstResponder = field

    let response = run(alert)
    let name = field.stringValue.trimmingCharacters(in: .whitespaces)
    switch response {
    case .alertFirstButtonReturn: return (name, .rename)
    case .alertSecondButtonReturn: return (name, .remove)
    default: return nil
    }
}

/// Tells the user an item with this name already exists.
func alertDuplicate(_ name: String) {
    let alert = NSAlert()
    alert.alertStyle = .warning
    alert.messageText = "“\(name)” already exists"
    alert.informativeText = "Item names must be unique."
    _ = run(alert)
}

/// Asks before resetting a streak — a misclick would kill the trophy.
/// length is the current streak ("4 months"), or "" when it's too young
/// to phrase that way. Cancel is the default; true means end it.
func confirmEndStreak(_ name: String, length: String) -> Bool {
    let alert = NSAlert()
    alert.messageText = length.isEmpty
        ? "End “\(name)” streak?"
        : "End your \(length) “\(name)” streak?"
    alert.addButton(withTitle: "Cancel")
    alert.addButton(withTitle: "End Streak")
    return run(alert) == .alertSecondButtonReturn
}

/// Confirm dialog for deletion; Cancel is the default. True means delete it.
func confirmRemove(_ name: String) -> Bool {
    let alert = NSAlert()
    alert.messageText = "Remove “\(name)”?"
    alert.addButton(withTitle: "Cancel")
    alert.addButton(withTitle: "Remove")
    return run(alert) == .alertSecondButtonReturn
}

private func textField() -> NSTextField {
    NSTextField(frame: NSRect(x: 0, y: 0, width: 230, height: 24))
}

/// Brings the app forward (it's an accessory app with no windows) and
/// runs the alert modally.
private func run(_ alert: NSAlert) -> NSApplication.ModalResponse {
    NSApp.activate(ignoringOtherApps: true)
    return alert.runModal()
}
